[#157] 글쓰기: 사진 촬영 기능 구현

jwonyLee opened the pull request:
## 개요

글쓰기> 사진 선택 화면에서 사진 촬영 기능을 구현했습니다.

## 작업 사항

- Info.plist > 사진 촬영 권한 추가
- 사진 촬영 셀을 위해 별도의 CenterIconCollectionCell 구현
- 사진 촬영 권한 허용 시, imagePickerController 표시
- 사진 촬영 권한 거부 시, 시스템 설정으로 이동하는 알럿 표시
- UIImagePickerController에서 사진 촬영 후 이미지를 편집할 때 중앙에서 이동할 수 없는 문제 해결
  - 이 부분은 아주 전 부터 있던 고질적인 문제인 거 같더라구요, 스택오버플로우 코드 참고했습니다.

## 예외 사항

- 시뮬레이터에서는 동작하지 않습니다. 
- 글쓰기 화면으로 이미지를 전달하는 로직은 현수님이 작업하신 풀 리퀘스트 #248 이 반영되고 난 후에 추가로 작업하려 합니다.

### Linked Issue

close #157

jwonyLee committed:
[#157] feat: 카메라 기능 구현
- 사용자 권한 요청
- 권한이 있으면 ImagePickerController를 이용해 사진 촬영
- 권한이 없으면 설정으로 이동하는 알럿 표시

jwonyLee committed:
[#157] feat: UIImagePickerController 에서 이미지를 편집할 때 중앙에서 이동할 수 없는 문제 해결

jwonyLee committed:
[#157] view: 사진 촬영 셀 변경
- CenterIconCollectionCell 구현

jwonyLee committed:
[#157] ref: 카메라 권한 요청 코드 수정
- imagePickerController를 present 함으로써 받는 권한 요청은, 사용자가 취소를 눌렀을 때도 이미 카메라 화면이 보여지고 있기 때문에 해당 부분은 변경

gustn3965 commented on ThingLog/ViewController/Photos/PhotosViewController+CollectionView.swift:
```diff
@@ -83,8 +83,23 @@ extension PhotosViewController: UICollectionViewDataSource {
         if indexPath.item == 0 {
-            // TODO: 카메라 기능 구현
-            return
+            switch AVCaptureDevice.authorizationStatus(for: .video) {
+            case .authorized:
+                present(imagePickerController, animated: true)
+            case .notDetermined:
+                AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
+                    guard let self = self else { return }
+                    if granted {
+                        self.present(self.imagePickerController, animated: true)
+                    }
+                }
```
requestAccess의 핸들러는 `The completion handler is called on an arbitrary dispatch queue` 호출될 수 있다고 하고, 실제로 디바이스에서 실행해보면 메인쓰레드에서 호출되어야한다고 에러가 나고 있습니다. main 쓰레드로 감싸면 될 것 같습니다.

gustn3965 commented on ThingLog/ViewController/Photos/PhotosViewController.swift:
```diff
@@ -306,3 +316,21 @@ extension PhotosViewController: PHPhotoLibraryChangeObserver {
+                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
+        picker.dismiss(animated: true)
+
+        guard let image = info[.editedImage] as? UIImage else {
+            print("\(#function): No edited image were found.")
+            return
+        }
+
+        // print out the image size as a test
+        print(image.size)
+        // TODO: WriteViewModel로 편집된 이미지 전달
+        coordinator?.back()
+    }
+}
```
문제점은 없지만, 추후에 이미지를 저장할 때, 사이즈를 조정할 필요가 있을 것 같습니다. 현재 아이폰 12 프로에서 찍고 보면 사이즈가 디바이스 사이즈보다 크게 나오는 것 같아서요 ~!

jwonyLee replied:
해당 부분은 이미지 전달을 구현할 때 resize 해서 전달할게요!

gustn3965 commented on ThingLog/View/CollectionViewCell/CenterIconCollectionCell.swift:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    private func setupView() {
+        backgroundColor = SwiftGenColors.primaryBackground.color
+        addSubview(iconView)
+
+        NSLayoutConstraint.activate([
+            iconView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            iconView.centerYAnchor.constraint(equalTo: centerYAnchor),
+            iconView.widthAnchor.constraint(equalToConstant: iconSize),
+            iconView.heightAnchor.constraint(equalTo: iconView.widthAnchor)
+        ])
+    }
+}
```
`contentView` 에 추가하는것이 좋을 것 같습니당. `centerX`, `Y` 도요 ~

gustn3965 reviewed (approved):
LGTM,  리뷰남기신거 반영하시고 머지하시면 될 것 같아요 ~

jwonyLee committed:
[#157] ref: 피드백 반영
- CenterIconCollectionCell contentView에 추가, 제약 설정
- AVCaptureDevice.requestAccess handler 안에서 present 할 때 메인 스레드에서 동작하게 변경

jwonyLee committed:
Merge branch 'develop' into feature/157-camera

jwonyLee commented:
피드백 반영했습니다~ 감사합니당
